block-layer/codecs/DAG-JSON.md

```diff
@@ -6,6 +6,17 @@ DAG-JSON supports the full [IPLD Data Model](../data-model-layer/data-model.md).
 
 ## Format
 
+### Serialization
+
+While it is unlikely de-serializers will enforce the following rules, codec implementors
+**SHOULD** do the following in order to ensure hashes consistently match for the same block data.
+
+* utf8 sort object keys
```

Comment:
"utf8 sort" isn't really specific enough. Do you mean the Unicode Collation with the DUCET table? With the roor collation? Or sorting by code points?

I think we should go for something like http://gibson042.github.io/canonicaljson-spec/. Where those things are clearly specified (though also there it took me a while to understand how they actually sort).

Reply:
I think we just sort lexicographically by encoded utf8. Note: we don't _currently_ bother normalizing anything first.

We need to handle that (for _all_ paths in _all_ formats) but that's a can of worms we've repeatedly opened and then re-shut).

Reply:
> I think we just sort lexicographically by encoded utf8.

What does lexicographically mean? Which locale?

Reply:
It's sorted using the _encoded_ (utf8) representation. That is, we just sort the bytes. See https://tools.ietf.org/html/rfc7049#section-3.9

Reply:
I see. Could we then change this to "sort UTF-8 encoded objects keys bytewise" or something similar?

Reply:
Let’s use the language [from wikipedia](https://en.m.wikipedia.org/wiki/UTF-8): “Sorting order: The chosen values of the leading bytes means that a list of UTF-8 strings can be sorted in code point order by sorting the corresponding byte sequences.”

Reply:
@mikeal sounds good to me. As long as there's a bit more information what "utf8 sort" is supposed to mean.

Reply:
I both like the language from wikipedia, but also support a flat, clear-speak statement like 

> That is, we just sort the bytes

... because it's true.  And character encodings are terrifying enough by their nature that clearspeak about the subject should be highly prized.

Reply:
“Sort by bytes” is going to be rather confusing to many developers since most of them don’t know that this is how utf-8 strings sort already, which may be what the keys are already encoded into. We don’t want to imply that it’s necessary to convert the string to binary in order to sort properly. We also need to worry about languages that default to ascii encoded strings (Python 2.x), because those won’t sort properly without being converted to utf-8.

I’m normally in favor of rather terse statements but I’m beginning to think this needs a bit more text than I’ve been giving it so far.

Reply:
> “Sort by bytes” is going to be rather confusing to many developers

I find that really clear. Look at the UTF-8 string byte by byte and compare it.

>  most of them don’t know that this is how utf-8 strings sort already

I don't understand what you mean by this. Perhaps I've been to deep into Unicode ordering/collation. I have no clue how utf-8 strings sort. I would need instructions, e.g. sort them by byte order. Or do proper collation with a locale. I don't see how there is a "default sort order".

Reply:
> I don't understand what you mean by this. Perhaps I've been to deep into Unicode ordering/collation. I have no clue how utf-8 strings sort. I would need instructions, e.g. sort them by byte order. Or do proper collation with a locale. I don't see how there is a "default sort order".

You don’t need to because most languages and/or standard libraries have sort functions that do this for you :) If the keys are already utf8, this is what people should use and they should **not** convert them to a binary object and do the sorting themselves ;)